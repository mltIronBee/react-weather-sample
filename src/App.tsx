import React from "react";
import { HomePage } from "@src/containers/pages/home";

function App(): JSX.Element {
	return <HomePage />;
}

export default App;
