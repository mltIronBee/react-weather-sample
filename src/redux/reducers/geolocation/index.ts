import { createSlice, PayloadAction } from "@reduxjs/toolkit";

export interface IGeolcationState {
	lat: number | null;
	lng: number | null;
}

const initialState: IGeolcationState = {
	lat: null,
	lng: null,
};

const geolocationSlice = createSlice({
	name: "geolocation",
	initialState,
	reducers: {
		getCurrentLocationSuccess(state, action: PayloadAction<IGeolcationState>) {
			state.lat = action.payload.lat;
			state.lng = action.payload.lng;
		},
	},
});

export const { getCurrentLocationSuccess } = geolocationSlice.actions;

export type GeolocationActions = ReturnType<typeof getCurrentLocationSuccess>;

export default geolocationSlice.reducer;
