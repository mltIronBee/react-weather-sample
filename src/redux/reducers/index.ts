import { combineReducers } from "@reduxjs/toolkit";
import weatherReducer, { IWeatherState, WeatherActions } from "@redux/reducers/weather";

export interface IAppState {
	weather: IWeatherState;
}

export type AppActions = WeatherActions;

export const rootReducer = combineReducers({ weather: weatherReducer });

export type RootReducer = ReturnType<typeof rootReducer>;
