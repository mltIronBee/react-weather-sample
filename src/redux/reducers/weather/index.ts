import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { AppThunk } from "@redux/store";
import { getWeatherForecast } from "@services/weather";

export interface IDailyForecast {
	readonly minTemperature: number;
	readonly maxTemperature: number;
	readonly dayOfWeek: string;
}

export interface IWeatherState {
	readonly isLoading: boolean;
	readonly errorMessage?: string;
	readonly forecast: IDailyForecast[];
}

const initialState: IWeatherState = {
	forecast: [],
	isLoading: false,
	errorMessage: "",
};

const weatherSlice = createSlice({
	name: "weather",
	initialState,
	reducers: {
		getForecastStart(state) {
			state.isLoading = true;
		},
		getForecastSuccess(state, action: PayloadAction<IDailyForecast[]>) {
			state.isLoading = false;
			state.errorMessage = "";
			state.forecast = action.payload;
		},
		getForecastError(state, action: PayloadAction<string>) {
			state.isLoading = false;
			state.errorMessage = action.payload;
		},
	},
});

export const { getForecastStart, getForecastError, getForecastSuccess } = weatherSlice.actions;

export type WeatherActions =
	| ReturnType<typeof getForecastStart>
	| ReturnType<typeof getForecastError>
	| ReturnType<typeof getForecastSuccess>;

export const fetchWeatherForecast = (city: string): AppThunk<Promise<void>> => async (dispatch) => {
	dispatch(getForecastStart());

	try {
		const forecast = await getWeatherForecast(city);

		dispatch(getForecastSuccess(forecast));
	} catch (error) {
		dispatch(getForecastError(error.toString()));
	}
};

export default weatherSlice.reducer;
