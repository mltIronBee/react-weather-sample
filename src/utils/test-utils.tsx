import React from "react";
import { render, RenderOptions, RenderResult } from "@testing-library/react";
import { ThemeProvider } from "@material-ui/core";
import { GenerateId } from "jss";
import { StylesProvider } from "@material-ui/core/styles";
import theme from "@src/theme";

const generateClassName: GenerateId = (rule, sheet) => `${sheet!.options.classNamePrefix}-${rule.key}`;

const AllTheProviders: React.FC = ({ children }) => (
	<StylesProvider generateClassName={generateClassName}>
		<ThemeProvider theme={theme}>{children}</ThemeProvider>
	</StylesProvider>
);

const customRender = (ui: React.ReactElement, options?: Omit<RenderOptions, "queries">): RenderResult =>
	render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
// eslint-disable-next-line import/export
export * from "@testing-library/react";

// override render method
// eslint-disable-next-line import/export
export { customRender as render };
