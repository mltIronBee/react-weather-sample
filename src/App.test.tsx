import React from "react";
import { Provider } from "react-redux";
import { getDefaultMiddleware } from "@reduxjs/toolkit";
import configureStore from "redux-mock-store";
import { render, screen, waitFor } from "@utils/test-utils";
import App from "@src/App";
import { IAppState } from "@redux/reducers";
import "@testing-library/jest-dom/extend-expect";

describe("App container", () => {
	const store = configureStore<IAppState>(getDefaultMiddleware())({
		weather: {
			isLoading: false,
			forecast: [],
			errorMessage: "",
		},
	});

	it("Should render correctly", () => {
		const { container } = render(
			<Provider store={store}>
				<App />
			</Provider>,
		);

		expect(container.firstChild).toMatchSnapshot();
	});
	it("Should render title", async () => {
		render(
			<Provider store={store}>
				<App />
			</Provider>,
		);
		const titleElement = await waitFor(() => screen.findByText(/React weather forecast/i));

		expect(titleElement).toBeInTheDocument();
	});
});
