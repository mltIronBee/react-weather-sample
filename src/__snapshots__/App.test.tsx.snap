// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App container Should render correctly 1`] = `
<div
  class="makeStyles-container"
>
  <div
    class="makeStyles-loadingWrapper"
  >
    <div
      class="makeStyles-loadingIndicatorContainer"
    >
      <div
        class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
        role="progressbar"
        style="width: 40px; height: 40px;"
      >
        <svg
          class="MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </div>
      <p
        class="MuiTypography-root MuiTypography-body1"
      >
        Loading...
      </p>
    </div>
    <div
      class="makeStyles-container"
    />
  </div>
</div>
`;
